### PR TITLE
New quest execution system

### PR DESCRIPTION
**Pull request by ZakShearman**
The aim of this is PR is to make quest execution use a builder to massively clean up how quest execution works and make it easier to document.

- [x] Old methods deprecated and fit to work with new methods
- [x] Structure confirmed
- [x] Documented
- [x] Fully tested
- [x] Merge ready

**Seeing the new system**
To any developers that are reading this, please take a look at the existing QuestExecutor and ExternalQuestExecutor classes you already know to see the deprecated methods. From here you can see how the new methods will be built.

**Regarding deprecation**
We will keep the old methods and executors present for at least the next 3 months all marked as deprecated. After this period of time, the old methods will no longer be included in the plugin and the new system **MUST** be used or you will receive errors.

All feedback/code review from anyone is welcome related to this change. Thanks 💜.

### CODE REVIEW

File: src/main/java/io/github/battlepass/service/CheckHelper.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import lombok.experimental.UtilityClass;
+import org.jetbrains.annotations.Contract;
+
+@UtilityClass
+public class CheckHelper {
+
+    @Contract("null, _ -> fail")
+    public static void notNull(Object arg, String identifier) {
+        if (arg == null)
+            throw new IllegalArgumentException(identifier.concat(" cannot be null."));
```
Comment (alexwith) [inline src/main/java/io/github/battlepass/service/CheckHelper.java:12]: Please wrap code in {} to make it clear.

File: src/main/java/io/github/battlepass/service/CheckHelper.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    }
+
+    @Contract("!null, _ -> fail")
+    public static void isNull(Object arg, String message) {
+        if (arg != null)
+            throw new IllegalArgumentException(message);
```
Comment (alexwith) [inline src/main/java/io/github/battlepass/service/CheckHelper.java:18]: same here

File: src/main/java/io/github/battlepass/quests/workers/pipeline/steps/QuestValidationStep.java [hunk 3]
```diff
@@ -115,11 +116,11 @@ public boolean isQuestPrimarilyValid(User user, Quest quest, BigInteger progress
...
     }
 
-    public boolean isQuestValid(Player player, User user, Quest quest, BigInteger progress, boolean overrideUpdate) {
-        if (!this.isQuestPrimarilyValid(user, quest, progress, overrideUpdate)) {
+    public boolean isQuestValid(QuestExecution questExecution, Quest quest) {
+        if (!this.isQuestPrimarilyValid(questExecution.getUser(), quest, questExecution.getProgress(), questExecution.shouldOverrideUpdate())) {
             return false;
         }
-        String playerWorld = player.getWorld().getName();
+        String playerWorld = questExecution.getPlayer().getWorld().getName();
```
Comment (alexwith) [inline src/main/java/io/github/battlepass/quests/workers/pipeline/steps/QuestValidationStep.java:123]: Maybe pull "questExecution.getPlayer()" out to a local variable for clarity, nothing wrong with this though. And same with the instances of this above.

File: src/main/java/io/github/battlepass/quests/service/executor/QuestExecutionBuilderImpl.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    @Override
+    public QuestExecutionBuilder questResult(UnaryOperator<QuestResult> resultOperator) {
+        this.questResult = (ExecutableQuestResult) resultOperator.apply(new ExecutableQuestResult());
+        return this;
+    }
+
+    @Override
+    public QuestExecutionBuilder root(String root) {
+        if (this.questResult == null)
+            this.questResult = new ExecutableQuestResult();
```
Comment (alexwith) [inline src/main/java/io/github/battlepass/quests/service/executor/QuestExecutionBuilderImpl.java:66]: Please wrap this in {}, and all other instances of this. Once again personal preference, but just makes it more robust for future, and looks more structured in my opinion.

File: src/main/java/io/github/battlepass/quests/service/executor/QuestExecutionBuilderImpl.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+import java.math.BigInteger;
+import java.util.function.UnaryOperator;
+
+public class QuestExecutionBuilderImpl implements QuestExecutionBuilder {
+    private final QuestPipeline pipeline;
+    private Player player;
+
+    private final String questType;
+    private BigInteger progress;
+    private boolean overrideUpdate;
```
Comment (alexwith) [inline src/main/java/io/github/battlepass/quests/service/executor/QuestExecutionBuilderImpl.java:20]: Conventional practice to sort variables according to their scopes, so

private final
[...]
private final

private
[...]
private

File: src/main/java/io/github/battlepass/quests/service/executor/QuestExecutionBuilder.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    @NotNull
+    static QuestExecutionBuilder of(@NotNull QuestContainer container, @NotNull String questType) {
+        String buildType;
+        if (container instanceof ExternalQuestContainer) {
+            buildType = ((ExternalQuestContainer) container).getPrefix().concat(questType);
+        } else {
+            buildType = questType;
+        }
+        return new QuestExecutionBuilderImpl(container.getQuestPipeline(), buildType);
+    }
```
Comment (alexwith) [inline src/main/java/io/github/battlepass/quests/service/executor/QuestExecutionBuilder.java:33]: Maybe move this method to the top or bottom of the interface?

File: src/main/java/io/github/battlepass/quests/service/base/QuestContainer.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    }
+
+    protected QuestExecutionBuilder executionBuilder(String questType) {
+        return QuestExecutionBuilder.of(this, questType);
+    }
+
+    public QuestPipeline getQuestPipeline() {
+        return this.questPipeline;
+    }
+
```
Comment (alexwith) [inline src/main/java/io/github/battlepass/quests/service/base/QuestContainer.java:22]: remove this space :)

File: src/main/java/io/github/battlepass/logger/containers/LogContainer.java [hunk 8]
```diff
@@ -13,6 +14,10 @@ public static LogContainer of(String message, Player player) {
         return new BasicPlayerContainer(message, player);
     }
 
+    public static LogContainer of(QuestExecution questExecution) {
+        return new QuestExecutionContainer(questExecution);
+    }
```
Comment (alexwith) [inline src/main/java/io/github/battlepass/logger/containers/LogContainer.java:19]: This convention should be followed throughout the plugin.

**Review by alexwith — CHANGES_REQUESTED** [review]
It's been amazing to see you go from noob to a real good dude. (rhymes if you say it)

File: src/main/java/io/github/battlepass/logger/containers/LogContainer.java (hunk 8 above)
Comment (ZakShearman) [reply src/main/java/io/github/battlepass/logger/containers/LogContainer.java:19]: As in to change this part or change other existing parts of the plugin to match this format?

File: src/main/java/io/github/battlepass/quests/service/executor/QuestExecutionBuilder.java (hunk 6 above)
Comment (ZakShearman) [reply src/main/java/io/github/battlepass/quests/service/executor/QuestExecutionBuilder.java:33]: Moved to top

File: src/main/java/io/github/battlepass/quests/workers/pipeline/steps/QuestValidationStep.java (hunk 3 above)
Comment (ZakShearman) [reply src/main/java/io/github/battlepass/quests/workers/pipeline/steps/QuestValidationStep.java:123]: Created variables for Player for both in the QuestValidationStep

File: src/main/java/io/github/battlepass/logger/containers/LogContainer.java (hunk 8 above)
Comment (alexwith) [reply src/main/java/io/github/battlepass/logger/containers/LogContainer.java:19]: As in any method that constructs a class should be above the class' constructor. Or below, just consistent across the plugin.

Comment (ZakShearman) [reply src/main/java/io/github/battlepass/logger/containers/LogContainer.java:19]: Moved down under the constructor.

**Comment by ZakShearman** [comment]
Documentation on the wiki is not yet completed so for now please look at existing internal quests (2 aren't converted) and the javadocs

**Pull request merged**